envoyproxy/envoy
# AWS common: region handling cleanup

Commit Message: AWS common: region handling cleanup
Additional Description: Refactor region handling code as per https://github.com/envoyproxy/envoy/issues/31676
Risk Level: Low
Testing: Unit
Docs Changes: Yes, new region documentation describes region string retrieval process
Release Notes: Changelog
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] 31676
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

This PR uplifts some of the AWS common code to ensure regions are read consistently from environment and config/credentials file across extensions.

The following changes were made to accommodate this:

- Move environment variable/default code from credential_provider_impl.cc to utility.cc
- Move config/credentials file parsing code from credential_provider_impl.cc to utility.cc and make this generic to support both credentials and region use cases
- Create support functions in utility.cc to resolve a profile string and credentials/config file paths
- Update region_provider_impl to provide a region chain, that retrieves region string from environment/config/credentials file
- Update aws_request_signer extension to use the region chain if region is not explicit in envoy config, and make the regional optional in protobuf 
- Update grpc_credentials/aws_iam extension to use the region chain if region is not explicit in envoy config
